### PR TITLE
(DS) feat(TT): Improve TT architecture (Validated by AB testing)

### PR DESCRIPTION
**Pull request by lmontier-pass**
# PR template

image
=> last_check = model après le retour de PR

=> Voir [ici le détail sur mlflow](https://mlflow.passculture.team/#/experiments/38?searchFilter=&orderByKey=attributes.start_time&orderByAsc=false&startTime=ALL&lifecycleFilter=Active&modelVersionFilter=All+Runs&datasetsFilter=W10%3D)

### CODE REVIEW

File: jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py [hunk 1]
```diff
@@ -60,19 +63,18 @@ def __init__(
 
     def set_task(self, item_dataset=None):
         self.task = tfrs.tasks.Retrieval(
-            loss=tf.keras.losses.CategoricalCrossentropy(
+            tf.keras.losses.CategoricalCrossentropy(
```
Comment (cibrahim-pass) [inline jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py:66]: why not loss= ?

File: jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py [hunk 2]
```diff
@@ -139,19 +135,29 @@ def __init__(
 
         self._embedding_layers = {}
         for layer_name, layer_class in self.input_embedding_layers.items():
+            if isinstance(layer_class, NumericalFeatureProcessor):
+                training_data = self.data[layer_name]
+            else:
+                training_data = self.data[layer_name].unique()
             self._embedding_layers[layer_name] = layer_class.build_sequential_layer(
-                vocabulary=self.data[layer_name].unique()
+                vocabulary=training_data
```
Comment (cibrahim-pass) [inline jobs/ml_jobs/algo_training/two_towers_model/models/two_towers_model.py:143]: modify the vocabulary parameter to be clearer

File: jobs/ml_jobs/algo_training/two_towers_model/train.py [hunk 3]
```diff
@@ -62,13 +76,17 @@ def load_datasets(
...
     logger.info("Loading & processing datasets")
 
-    train_data = read_from_gcs(
-        storage_path=STORAGE_PATH, table_name=training_table_name
-    )[user_columns + item_columns].astype(str)
+    train_data = (
+        read_from_gcs(storage_path=STORAGE_PATH, table_name=training_table_name)[
+            user_columns + item_columns
+        ]
+        .drop_duplicates(subset=["user_id", "item_id"])
```
Comment (cibrahim-pass) [inline jobs/ml_jobs/algo_training/two_towers_model/train.py:83]: should we drop duplicates ? try training again without dropping or adding a feature with nuymber of interactions

Comment (lmontier-pass) [reply jobs/ml_jobs/algo_training/two_towers_model/train.py:83]: good point. I will create a task to readd the duplicates later on and compare it !

File: jobs/ml_jobs/algo_training/two_towers_model/train.py [hunk 4]
```diff
@@ -167,10 +194,10 @@ def log_mlflow_params(
 
 
 def train_two_tower_model(
-    item_dataset,
     train_dataset,
     validation_dataset,
     two_tower_model,
+    training_steps,
```
Comment (cibrahim-pass) [inline jobs/ml_jobs/algo_training/two_towers_model/train.py:200]: explain the epoch, shuffle, training steps and repeat logic in readme ?

**Review by lmontier-pass — COMMENTED** [review]
/gpt

**Review by cibrahim-pass — APPROVED** [review]
